chiselFrontend/src/main/scala/chisel3/Module.scala

```diff
@@ -0,0 +1,356 @@
+    val clockAndReset: Option[ClockAndReset] = Builder.currentClockAndReset
+    Builder.currentClockAndReset = None
+
+    // Execute the module, this has the following side effects:
+    //   - set currentModule
+    //   - unset readyForModuleConstr
+    //   - reset whenDepth to 0
+    //   - set currentClockAndReset
+    val module: T = bc  // bc is actually evaluated here
+
+    if (Builder.whenDepth != 0) {
+      throwException("Internal Error! when() scope depth is != 0, this should have been caught!")
+    }
+    if (Builder.readyForModuleConstr) {
+      throwException("Error: attempted to instantiate a Module, but nothing happened. " +
+                     "This is probably due to rewrapping a Module instance with Module()." +
+                     sourceInfo.makeMessage(" See " + _))
+    }
+    Builder.currentModule = parent // Back to parent!
+    Builder.whenDepth = whenDepth
+    Builder.currentClockAndReset = clockAndReset // Back to clock and reset scope
+
+    val component = module.generateComponent()
+    Builder.components += component
+
+    // Handle connections at enclosing scope
+    if(!Builder.currentModule.isEmpty) {
+      pushCommand(DefInstance(sourceInfo, module, component.ports))
+      module.initializeInParent(compileOptions)
+    }
+    module
+  }
+
+  /** Returns the implicit Clock */
+  def clock: Clock = Builder.forcedClock
+  /** Returns the implicit Reset */
+  def reset: Reset = Builder.forcedReset
+  /** Returns the current Module */
+  def currentModule: Option[BaseModule] = Builder.currentModule
+}
+
+package experimental {
+
+  object IO {
+    /** Constructs a port for the current Module
+      *
+      * This must wrap the datatype used to set the io field of any Module.
+      * i.e. All concrete modules must have defined io in this form:
+      * [lazy] val io[: io type] = IO(...[: io type])
+      *
+      * Items in [] are optional.
+      *
+      * The granted iodef must be a chisel type and not be bound to hardware.
+      *
+      * Also registers a Data as a port, also performing bindings. Cannot be called once ports are
+      * requested (so that all calls to ports will return the same information).
+      * Internal API.
+      */
+    def apply[T<:Data](iodef: T): T = {
+      val module = Module.currentModule.get // Impossible to fail
+      require(!module.isClosed, "Can't add more ports after module close")
+      requireIsChiselType(iodef, "io type")
+
+      // Clone the IO so we preserve immutability of data types
+      val iodefClone = try {
+        iodef.cloneTypeFull
+      } catch {
+        // For now this is going to be just a deprecation so we don't suddenly break everyone's code
+        case e: AutoClonetypeException =>
+          Builder.deprecated(e.getMessage, Some(s"${iodef.getClass}"))
+          iodef
+      }
+      module.bindIoInPlace(iodefClone)
+      iodefClone
+    }
+  }
+
+  object BaseModule {
```

Comment:
I don't think this object is exposed as an API even in experimental, it's just chisel3.core. Can this be moved into internal?

Reply:
It's exposed in the chisel3 package object in `experimental`

Reply:
The type appears to be exposed, but not the object
https://github.com/freechipsproject/chisel3/blob/master/src/main/scala/chisel3/package.scala#L442

Reply:
Good point @ducky64. That subtlety escaped me.